src/jit/morph.cpp

```diff
@@ -12518,12 +12521,12 @@ GenTreePtr Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* mac)
                     // castType is larger or the same as op1's type
                     // then we can discard the cast.
 
-                    if (varTypeIsSmall(castType) && (genTypeSize(castType) >= genTypeSize(op1->TypeGet())))
+                    if (varTypeIsSmall(castType) && (genTypeSize(castType) >= genTypeSize(effectiveOp1->TypeGet())))
                     {
                         tree->gtOp.gtOp2 = op2 = op2->gtCast.CastOp();
                     }
                 }
-                else if (op2->OperIsCompare() && varTypeIsByte(op1->TypeGet()))
+                else if (op2->OperIsCompare() && varTypeIsByte(effectiveOp1->TypeGet()))
```

Comment:
This is another weird transform that's specific to x86/x64 and should be done in lowering. At least this case can be reproduced and verified that it works correctly.